using float type to do the computation of norm reduce for cpu half and bfloat16 dtype

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95166

As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10

XiaobingSuper committed:
using accumulate type to do the computation of norm reduce for cpu path

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/cpu/ReduceNormKernel.cpp:
```diff
@@ -0,0 +1,138 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
```
Nit. When moving things around, it's good to provide a diff of the two files, as otherwise it's difficult to figure out what's different between the two.

peterbell10 replied:
Also don't just move things around for the sake of it. CUDA files are split up because they take much longer to compile so smaller files is better for parallelisation, whereas CPU files are larger to amoritize the cost of pre-processing.

lezcano commented on aten/src/ATen/native/cpu/ReduceNormKernel.cpp:
```diff
@@ -0,0 +1,138 @@
+              });
+        });
+  } else {
+    if (iter.dtype(0) == kHalf) {
+      return norm_kernel_cpu_impl<at::Half, float>(iter, val);
+    } else if (iter.dtype(0) == kBFloat16) {
+      return norm_kernel_cpu_impl<at::BFloat16, float>(iter, val);
+    }
```
You can copy the logic for how CUDA does it, as it's able to fuse this operation and a casting operation in one in some cases.

XiaobingSuper replied:
Yes, using the same logic as cuda.

lezcano reviewed (commented):
The change looks good to me. Can you add specific tests for this change? I tested the changes in `test_linalg.py` in my local branch and they pass.

peterbell10 commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -192,9 +192,11 @@ struct MeanOps {
+template <
+    typename scalar_t,
+    typename acc_t = scalar_t,
+    typename out_t = scalar_t>
```
If we're going to have defaults should it not be `out_t = acc_t` for backward compatibility?

XiaobingSuper replied:
For the Norm ops, I think it is strange to set `out_t = acc_t`, from the design at https://github.com/pytorch/pytorch/blob/92e03cd583c027a4100a13682cf65771b80569da/aten/src/ATen/native/cpu/Reduce.h#L166-L170, we hope the `project ` return the required output.

peterbell10 replied:
You have changed the meaning of `NormOps<scalar_t, acc_t>` so it now outputs `scalar_t` instead of `acc_t` which is wrong for complex types where the norm is always real.

XiaobingSuper replied:
Thanks for the explanation, changed the default value of `out_t ` to `acc_t `.

peterbell10 commented on aten/src/ATen/native/cpu/Reduce.h:
```diff
@@ -115,41 +115,57 @@ static inline void vectorized_outer_reduction(char** data, int64_t inner_stride,
-static void set_result(const int index, const res_t result, const TensorIteratorBase &iter, const int num_outputs) {
-  // static_assert(std::is_same<res_t, typename traits::arg2_t>::value, "data types must match");
+template <typename res_t>
+static void set_result(
```
Is this re-formatting done to appease the linter, or is it something your adding? It muddies the diff and subsequent git blame, so I'd prefer you didn't.

XiaobingSuper replied:
Yes, I re-formatted the code, and I reverted it now.

peterbell10 commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -237,140 +237,6 @@ static void prod_kernel_impl(TensorIterator& iter) {
-    // If we can vectorize over the last dimension and the dtype
-    // of the output is the same as that of the input,
-    // then we go through the vectorised path.
-    if (is_reduce_lastdim(iter) &&
```
What happened to this code?

lezcano replied:
It's up there in the new file. The only thing is that the comment is gone for no apparent reason

peterbell10 replied:
Ah, I was expecting to see it next to `NormTwoOps` but they've been split apart.

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 reviewed (approved):
LGTM. Please add UTs to cover the changes. Need to make sure the tests go into the scalar path.

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -138,7 +138,8 @@ void TestLoadOfAddsWithCopy(DeprecatedTypeProperties& type) {
             << " ms" << std::endl;
-  ASSERT_EQ_RESOLVED(norm(100000 * d).item<double>(), norm(r).item<double>());
+  
+  ASSERT_EQ_RESOLVED(norm(1000 * d).item<double>(), norm(r).item<double>());
```
Why changed to a smalled size?  see 
```
>>> a=torch.ones(3, 4).half().cuda()
>>> b=a*100000
>>> b
tensor([[inf, inf, inf, inf],
        [inf, inf, inf, inf],
        [inf, inf, inf, inf]],  dtype=torch.float16)
>>> r = torch.zeros(3, 4).half()
>>> for i in range(100000):
...     r = torch.add(r, a)
...
>>> r
tensor([[2048., 2048., 2048., 2048.],
        [2048., 2048., 2048., 2048.],
        [2048., 2048., 2048., 2048.]], dtype=torch.float16)
```
If we use float as accumulate type, the `r` input can get a `7096 number`, but the `b` input gets an `inf` number. I checked cuda path, this case is also failed for such a big input.

peterbell10 commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -250,55 +250,50 @@ inline void norm_two_reduce_step(Vectorized<float>& acc_fvec, Vectorized<BFloat1
-      );
-    });
-  } else if (val == 2) {
+  if (val == static_cast<double>(2) && is_reduce_lastdim(iter) &&
```
NIT:
```suggestion
  if (val == 2.0 && is_reduce_lastdim(iter) &&
```

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented:
> The change looks good to me. Can you add specific tests for this change? I tested the changes in `test_linalg.py` in my local branch and they pass.

Yes, I added a new test case to cover it:

```
test_norm_bfloat16_and_half_cpu_bfloat16 (test_linalg.TestLinalgCPU) ... FAIL
test_norm_bfloat16_and_half_cpu_float16 (test_linalg.TestLinalgCPU) ... FAIL

======================================================================
FAIL: test_norm_bfloat16_and_half_cpu_bfloat16 (test_linalg.TestLinalgCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/xiaobing/pytorch-offical/torch/testing/_internal/common_device_type.py", line 401, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/xiaobing/pytorch-offical/test/test_linalg.py", line 1203, in test_norm_bfloat16_and_half
    run_test_case((S,) , ord, keepdim, )
  File "/home/xiaobing/pytorch-offical/test/test_linalg.py", line 1199, in run_test_case
    self.assertEqual(result_ref, result, msg=msg)
  File "/home/xiaobing/pytorch-offical/torch/testing/_internal/common_utils.py", line 2967, in assertEqual
    raise error_metas[0].to_error(
AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 1 (100.0%)
Greatest absolute difference: 1792.0 at index (0,) (up to 1e-05 allowed)
Greatest relative difference: 7.0 at index (0,) (up to 0.016 allowed)
input_size=(2049,), ord=0, keepdim=True, dtype=torch.bfloat16

======================================================================
FAIL: test_norm_bfloat16_and_half_cpu_float16 (test_linalg.TestLinalgCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/xiaobing/pytorch-offical/torch/testing/_internal/common_device_type.py", line 401, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/xiaobing/pytorch-offical/test/test_linalg.py", line 1203, in test_norm_bfloat16_and_half
    run_test_case((S,) , ord, keepdim, )
  File "/home/xiaobing/pytorch-offical/test/test_linalg.py", line 1199, in run_test_case
    self.assertEqual(result_ref, result, msg=msg)
  File "/home/xiaobing/pytorch-offical/torch/testing/_internal/common_utils.py", line 2967, in assertEqual
    raise error_metas[0].to_error(
AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 1 (100.0%)
Greatest absolute difference: 0.015625 at index (0,) (up to 1e-05 allowed)
Greatest relative difference: 0.0012315270935960591 at index (0,) (up to 0.001 allowed)
input_size=(2049,), ord=3, keepdim=True, dtype=torch.float16

----------------------------------------------------------------------
Ran 2 tests in 0.015s

FAILED (failures=2)
```

XiaobingSuper commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -250,55 +250,50 @@ inline void norm_two_reduce_step(Vectorized<float>& acc_fvec, Vectorized<BFloat1
-      );
-    });
-  } else if (val == 2) {
+  if (val == static_cast<double>(2) && is_reduce_lastdim(iter) &&
```
Changed.

XiaobingSuper commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -250,55 +250,50 @@ inline void norm_two_reduce_step(Vectorized<float>& acc_fvec, Vectorized<BFloat1
+    binary_kernel_reduce(iter, NormTwoOps<scalar_t, acc_t, out_t>(), acc_t(0));
+  } else if (val == static_cast<double>(INFINITY)) {
+    binary_kernel_reduce(iter, AbsMaxOps<scalar_t, acc_t, out_t>(), acc_t(0));
+  } else if (val == static_cast<double>(-INFINITY)) {
```
Changed.

lezcano commented:
What about this point https://github.com/pytorch/pytorch/pull/95166#discussion_r1112188747?

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented:
@lezcano @peterbell10 , there have a failed case for MPS backend(compare CPU path with MPS path):
```
FAIL [0.014s]: test_output_match_nn_functional_pairwise_distance_cpu_float16 (__main__.TestConsistencyCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_4233535948/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 401, in instantiated_test
    result = test(self, **param_kwargs)
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_4233535948/lib/python3.9/site-packages/torch/testing/_internal/common_device_type.py", line 851, in test_wrapper
    return test(*args, **kwargs)
  File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/test_mps.py", line 9912, in test_output_match
    raise e
  File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/test_mps.py", line 9909, in test_output_match
    self.assertEqual(cpu_grad_inputs, mps_grad_inputs, atol=atol, rtol=rtol)
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_4233535948/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2967, in assertEqual
    raise error_metas[0].to_error(
AssertionError: Tensor-likes are not close!

Mismatched elements: 2 / 3 (66.7%)
Greatest absolute difference: 0.000732421875 at index (0,) (up to 1e-05 allowed)
Greatest relative difference: 0.0025020850708924102 at index (0,) (up to 0.001 allowed)
```
it seems MPS backend doesn't use float as accumulate  dtype for half input, see :https://github.com/pytorch/pytorch/blob/cf6e078c34767b4356663ab93b07efc7c984ac90/aten/src/ATen/native/mps/operations/ReduceOps.mm#L439, I think the norm MPS path should add dtype cast if input type is half or bfloat16 as https://github.com/pytorch/pytorch/blob/cf6e078c34767b4356663ab93b07efc7c984ac90/aten/src/ATen/native/mps/operations/ReduceOps.mm#L133.

Could you help cc the maintainer of MPS backend to help double-check it? if MPS also needs to use float as accumulate dtype for half or bfloat16 input, do we need to make changes or temporarily disable this test case at this PR? Thanks!

lezcano commented:
For now just disable the test on MPS. 
cc @kulinseth

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 commented on test/test_mps.py:
```diff
@@ -9566,7 +9566,9 @@ class TestConsistency(TestCaseMPS):
-        'nn.functional.pairwise_distance': ['f16', 'f32'],
+        # TODO: add f16 test case after solve the accuracy issue,
+        # see https://github.com/pytorch/pytorch/pull/95166#issuecomment-1439359181.
+        'nn.functional.pairwise_distance': ['f32'],
```
This list is for backwards. You need to modify the list above to skip forwards.

XiaobingSuper replied:
done.

XiaobingSuper committed:
Update on "using float type to do the computation of norm reduce for cpu half and bfloat16 dtype"


As the title, we should use a higher dtype to compute norm reduce for half and bfloat1 dtype.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
